### PR TITLE
Added predict_ensemble to all model classes

### PR DESCRIPTION
**Pull request by jsschreck**

### CODE REVIEW

File: evml/keras/models.py [hunk 1]
```diff
@@ -195,79 +267,251 @@ def load_model(cls, conf):
...
         _batch_size = self.batch_size if batch_size is None else batch_size
         y_out = self.model.predict(x, batch_size=_batch_size)
+        if scaler:
+            if len(y_out.shape) == 1:
+                y_out = np.expand_dims(y_out, 1)
+            y_out = scaler.inverse_transform(y_out)
         return y_out
+    
+    def predict_ensemble(self, x, weight_locations, batch_size=None, scaler=None, num_outputs=1):
+        num_models = len(weight_locations)
```
Comment (djgagne) [inline evml/keras/models.py:323]: Can you add a docstring for predict_ensemble? What is weight_locations supposed to be?

File: applications/train_classifier_ptype.py [hunk 2]
```diff
@@ -16,7 +16,7 @@
 from argparse import ArgumentParser
 
 from ptype.callbacks import MetricsCallback
-from ptype.data import load_ptype_data_day, preprocess_data
+from ptype.data import load_ptype_uq, preprocess_data
```
[the remark this answers is not in the document]
Comment (jsschreck) [reply applications/train_classifier_ptype.py:19]: I had the load_ptype_uq here but moved it to ptype recently :)

[the remark this answers is not in the document]
Comment (jsschreck) [reply applications/train_classifier_ptype.py:19]: One thing I could do is try to import ptype, except import error, then pip install ptype? The training scripts still depend on the models in other ways (such as how we split the data, preprocess, etc).

File: evml/keras/models.py (hunk 1 above)
Comment (jsschreck) [reply evml/keras/models.py:323]: Done! Its a list of paths to saved weights.

File: evml/keras/models.py [hunk 3]
```diff
@@ -136,8 +132,52 @@ def build_neural_network(self, inputs, outputs):
...
+        elif self.metrics == "mse":
+            metrics = self.mse
+        else:
+            metrics = None
+            
+        self.model.compile(
+            optimizer=self.optimizer_obj,
+            loss=self.loss,
+            loss_weights=self.loss_weights,
+            metrics=metrics,
```
Comment (charlie-becker) [inline evml/keras/models.py:165]: The `loss_weights` argument here is not the same as `class_weights` in the `.fit()` method which I think is your intention.

File: evml/keras/models.py [hunk 4]
```diff
@@ -983,9 +1157,33 @@ def predict_monte_carlo(self, x, mc_forward_passes=10, batch_size=None):
...
+        for i, weight_location in enumerate(weight_locations[1:]):
+            model_instance = self.model
+            model_instance.load_weights(weight_location)
+            y_prob = model_instance.predict(x, batch_size=batch_size)
+            predictions[i + 1] = y_prob
+
+        return predictions
 
     def compute_uncertainties(self, y_pred, num_classes=4):
         return calc_prob_uncertainty(y_pred, num_classes=num_classes)
```
Comment (charlie-becker) [inline evml/keras/models.py:1189]: In the other model classes, I think the `calc_prob_uncertainties()` is a method within the class, and here it is a function outside the class. It would be better if they were consistent.

**Review by charlie-becker — CHANGES_REQUESTED** [review]
I made some comments regarding the categorical evidential model which I've been using with p-type.

File: evml/keras/models.py [hunk 5]
```diff
@@ -523,184 +860,19 @@ def predict_dist_params(self, x, y_scaler=None, batch_size=None):
...
-        return mu, var
-
+        return super().predict_monte_carlo(x_test, y_test, forward_passes, scaler=scaler, batch_size=batch_size, num_outputs=3)
 
+    
 class CategoricalDNN(object):
-
     """
     A Dense Neural Network Model that can support arbitrary numbers of hidden layers.
     Attributes:
```
[the remark this answers is not in the document]
Comment (jsschreck) [reply evml/keras/models.py:878]: I am going to merge this PR soon and open a new one focused on the Categorical model! Hopefully it should be far fewer files to look at compared to this one.

File: evml/keras/models.py (hunk 3 above)
Comment (jsschreck) [reply evml/keras/models.py:165]: I will test again to see if this actually works (despite keras saying it does); if not it will be removed. But I will add class_weights as an option for when the model is a standard classifier and not evidential (we could enable the experimental class weights I put into the loss, but I cannot confirm that is really correct ...)

File: applications/train_classifier_ptype.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (djgagne) [reply applications/train_classifier_ptype.py]: We can do optional dependencies in the pyproject.toml file, so I would suggest throwing an exception on the import and adding some helpful error message to install ptype or run the pip install including optional dependencies.

**Review by djgagne — APPROVED** [review]
We are going to merge in the current changes and address Charlie's categorical comments and remaining refactoring issues in another pull request.
